(PUP-8984) Add --format=json option to `puppet parser validate`

scotje opened the pull request:
Tools like the Puppet Development Kit which invoke `puppet parser validate` on behalf of the user need a consistent, machine-readable output format so that they don't have to rely on complex and brittle regular expressions matching output that may change between Puppet releases.

DavidS commented:
Works for me, but the `--render-as` vs `--format` is really jarring:

```
david@davids:~/git/puppetlabs-ntp$ bundle exec puppet parser validate --help


USAGE: puppet parser validate [--format json] [<manifest>] [<manifest> ...]

This action validates Puppet DSL syntax without compiling a catalog or
syncing any resources. If no manifest files are provided, it will
validate the default site manifest.

RETURNS: Nothing, or the first syntax error encountered.

OPTIONS:
  --render-as FORMAT             - The rendering format to use.
  --verbose                      - Whether to log verbosely.
  --debug                        - Whether to log debug information.
  --format json                  - Format output as JSON instead of default.

See 'puppet help parser' or 'man puppet-parser' for full help.
david@davids:~/git/puppetlabs-ntp$ bundle exec puppet parser validate --format json manifests/init.pp 
{
  "manifests/init.pp": {
    "issue_code": "SYNTAX_ERROR",
    "message": "Syntax error at ')'",
    "full_message": "Could not parse for environment production: Syntax error at ')' (file: /home/david/git/puppetlabs-ntp/manifests/init.pp, line: 286, column: 3)",
    "file": "/home/david/git/puppetlabs-ntp/manifests/init.pp",
    "line": 286,
    "pos": 3,
    "environment": "production",
    "node": null,
    "arguments": null
  }
}
david@davids:~/git/puppetlabs-ntp$
```

I've played around a bit with the face API, and I came up with this:

```diff
david@davids:~/git/puppet$ git diff
diff --git a/lib/puppet/face/parser.rb b/lib/puppet/face/parser.rb
index f83dda254..e1f37482c 100644
--- a/lib/puppet/face/parser.rb
+++ b/lib/puppet/face/parser.rb
@@ -78,11 +78,20 @@ Puppet::Face.define(:parser, '0.0.1') do
           parse_errors.each { |_, error| Puppet.log_exception(error) }
         end
 
-        exit(1)
+        # exit(1)
       end
 
-      nil
+      parse_errors
     end
+
+    when_rendering :console do |parse_errors|
+      parse_errors.each {|p| puts p}
+    end
+
+    when_rendering :json do |parse_errors|
+      puts JSON.pretty_generate(parse_errors.map { |file, error| [file, error.to_h] }.to_h)
+    end
+
   end
 
 
david@davids:~/git/puppet$ 
```

results in this:

```
david@davids:~/git/puppetlabs-ntp$ bundle exec puppet parser validate --render-as json manifests/init.pp 
Error: Could not parse for environment production: Syntax error at ')' (file: /home/david/git/puppetlabs-ntp/manifests/init.pp, line: 286, column: 3)
{
  "manifests/init.pp": {
    "issue_code": "SYNTAX_ERROR",
    "message": "Syntax error at ')'",
    "full_message": "Could not parse for environment production: Syntax error at ')' (file: /home/david/git/puppetlabs-ntp/manifests/init.pp, line: 286, column: 3)",
    "file": "/home/david/git/puppetlabs-ntp/manifests/init.pp",
    "line": 286,
    "pos": 3,
    "environment": "production",
    "node": null,
    "arguments": null
  }
}
null
david@davids:~/git/puppetlabs-ntp$ 
```

puppetcla commented:
CLA signed by all contributors.

scotje commented:
@DavidS Ah, thanks for your help deciphering the face API. The other kind of annoying thing is that `puppet parser dump` implements it's own `--format` option and also ignores `--render-as`, so it's going to be inconsistent one way or another. :(

scotje commented:
Re-implemented with `--render-as` support which also helped clean up the spec tests. (Although I did have to dig around for a while to figure out how to invoke the output renderers from unit tests.)

hlindberg commented on lib/puppet/error.rb:
```diff
@@ -71,6 +71,20 @@ def to_s
+        :pos => pos,
+        :environment => environment,
+        :node => node,
+        :arguments => arguments,
```
I think this requires a different treatment since `arguments` can be any data type. If this error class instead complies with the protocol then you can do this:
```
      io = StringIO.new
      writer = Puppet::Pops::Serialization::JSON::Writer.new(io)
      serializer = Puppet::Pops::Serialization::Serializer.new(writer)
      serializer.write(data)
      serializer.finish
      io.string
```
to get the JSON string

hlindberg replied:
Ping @thallgren what options are needed to ensure any Ruby object gets a resonable representation in JSON when using the serializer. (As I recall it, arguments to issues can be any kind of Ruby Object).
Also wondering, if the logic in this face could use the "logging to json" support we added to support catalog preview. (Obviously it could set up logging the result to a `.json` tmp file and then render that to the console).

thallgren replied:
I think the simplest way to resolve this would be to use the `Puppet::Pops::Serialization::ToDataConverter` to ensure that the `arguments` value is pure Data (i.e. types supported by JSON). That conversion will be a no-op for most types of arguments. The options used should probably be an empty hash (I'm making the assumption that it's not very useful to convert rich-data with "\_\_ptype__" keywords).

hlindberg replied:
I don't think that works since some issues will send a Ruby `Class` IIRC - so there will be non-data, and it needs to just use `to_s` on such things. Didn't we do this for the .json file logger in Puppet? It must do the right thing because it works... Ping @thallgren

thallgren replied:
The `ToDataConverter` will use `to_s` on things that it doesn't recognize unless rich_data is enabled.

The .json file logger simply does this.
```
@file.puts(Puppet::Util::Json.dump(msg.to_structured_hash))
```
and to_structured_hash doesn't output arguments. It does:
```
  def to_structured_hash
    hash = {
      'level' => @level,
      'message' => @message,
      'source' => @source,
      'tags' => @tags.to_a,
      'time' => @time.iso8601(9),
    }
    %w(file line pos issue_code environment node backtrace).each do |name|
      attr_name = "@#{name}"
      hash[name] = instance_variable_get(attr_name) if instance_variable_defined?(attr_name)
    end
    hash
  end
```

scotje commented:
@hlindberg @thallgren So if I understand correctly, there are two things to address here:

1. There may already be code somewhere to turn a `Puppet::ParseErrorWithIssue` (or maybe just `Puppet::ParseError`) into a hash and that should be used instead of adding a new method to the class. @hlindberg can you point me in the general direction of where that code might be?

1. If included, the `@arguments` property needs special treatment to ensure it is serialized properly into JSON. I don't think we actually care about the value of the `@arguments` for our use case, so if there is an approach similar to the `to_structured_hash` method that @thallgren showed that simply excludes that property, I would be fine with that. But happy to implement the more robust serialization for that property as discussed.

melissa commented:
@hlindberg @thallgren  could you take another look at this. It would be good to move this forward.

The rich data support in puppet 6 will impact this implementation.

joshcooper commented:
Seems like we can just do:

```ruby
data = errors.map { |file, error| [file, error.to_h] }.to_h)
puts Puppet::Pops::Serialization::ToDataConverter.convert(data, rich_data: false)
```

I think the converter's default options "do the right thing" for the simple case.

scotje commented:
Updated PR based on feedback. Specifically I implemented @joshcooper 's suggested simplification.

Further research revealed that the `Puppet.log_exception` method has a special case for logging `ParseErrorWithIssue` that does it's own hash serialization of the class before creating a log event object out of it. https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/logging.rb#L53-L68

A further improvement might be to have that method call `to_h` on the `ParseErrorWithIssue` instance and then merge in the log specific attributes before creating the Log object. This could reduce the redundancy between the two approaches right now.

Another approach might be to extend the `:console` log destination to support multiple formats, for this particular feature it seems inefficient to be forced to write to a temp file (and then read it back so it can be output to STDOUT) just to get the logger's built-in JSON support.

melissa commented:
It looks like the last few things we need to figure out on this ticket is to address Henrik's comment on parsing `@arguments`. It looks like `@environment` and `@node` should also have some special treatment so that we only return the name of the object, not the object itself, depending on what gets passed in.

@kris-bosland will move this forward

hlindberg commented:
Note that validation result can contain issues of warning type as well

kris-bosland commented:
@scotje , can you remove the :arguments entry, and get the name or other suitable .to_s result for the :environment and :node entries?  Thanks!

scotje commented:
@melissa @kris-bosland My understanding was that the use of `ToDataConverter` would handle ensuring things are serialized to JSON-safe values. It doesn't seem correct to coerce things to strings in the `.to_h` method since things other than this might be interested in that output and would expect things to still be their original classes/types. As long as `ToDataConverter` is calling `to_s` on things it doesn't recognize we should be good?

I can delete the `arguments` key out of the hash right before it is serialized to JSON however, since we really don't care about that and it's probably the most likely thing to not serialize well.

scotje commented:
> I can delete the `arguments` key out of the hash right before it is serialized to JSON however, since we really don't care about that and it's probably the most likely thing to not serialize well.

I'm going to go ahead and have the face filter out `arguments`, `environment`, and `node` because none of those are really relevant to the `pupper parser validate` use case.

kris-bosland commented:
Thanks @scotje .  The reason to filter or remove those entries is because they can contain arbitrarily large data structures and those data structures don't contribute much to solving the users issue.  The name of the environment or node may be helpful, however.

scotje commented:
@kris-bosland I've removed `arguments` from the return value of `to_h` and explicitly called `to_s` on the `environment` and `node` values. I've also updated the Validate face to filter out all three of those keys before rendering it's output to JSON format since they are unlikely to be useful. We can always add one or more of them back in later should they prove useful.

Also rebased on master and updating tests to use `rspec-mocks`, etc.
